core/dbt/context/common.py

```diff
@@ -211,16 +211,22 @@ def search_package_name(self):
         return self.model.package_name
 
     def add_provider_functions(self, context):
-        context['ref'] = self.provider.ref(
-            self.db_wrapper, self.model, self.config, self.manifest
+        # Generate the builtin functions
+        builtins = dict(
+            ref=self.provider.ref(
+                self.db_wrapper, self.model, self.config, self.manifest),
+            source=self.provider.source(
+                self.db_wrapper, self.model, self.config, self.manifest),
+            config=self.provider.Config(
+                self.model, self.source_config),
+            execute=self.provider.execute
         )
-        context['source'] = self.provider.source(
-            self.db_wrapper, self.model, self.config, self.manifest
-        )
-        context['config'] = self.provider.Config(
-            self.model, self.source_config
-        )
-        context['execute'] = self.provider.execute
+        # Install them at .builtins
+        context['builtins'] = builtins
+        # Install each of them directly in case they're not
+        # clobbered by a macro.
+        for builtin in builtins:
+            context[builtin] = builtins[builtin]
```

Comment:
This is equivalent to `context.update(builtins)`, right? I find that a bit clearer, IMO the only reason to use a loop here is if you need to update without replacement (and in that case, I'd iterate over `.items()`).

Reply:
👍